common/json.c

```diff
@@ -216,3 +216,54 @@ jsmntok_t *json_parse_input(const char *input, int len, bool *valid)
 
 	return toks;
 }
+
+void json_tok_print(const char *buffer, const jsmntok_t *params)
+{
+	const jsmntok_t *first = params;
+	const jsmntok_t *last = json_next(params);
+	printf("size: %d, count: %ld\n", params->size, last - first);
+	while (first != last) {
+		printf("%ld. %.*s, %d\n", first - params,
+		        first->end - first->start, buffer + first->start,
+			first->type);
+		first++;
+	}
+	printf("\n");
+}
+
+jsmntok_t *json_tok_copy(const tal_t *ctx, const jsmntok_t *tok)
+{
+	const jsmntok_t *first = tok;
+	const jsmntok_t *last = json_next(tok);
+
+	if (!tok)
+		return NULL;
+	jsmntok_t *arr = tal_arr(ctx, jsmntok_t, last - first);
+	jsmntok_t *dest = arr;
+	while (first != last)
+		*dest++ = *first++;
+
+	return arr;
+}
```

Comment:
I think this is *almost* a 1-liner?

```
   if (!tok)
        return NULL;
   return tal_dup_arr(ctx, jsmntok_t, tok, json_next(tok) - tok, 0);
```

Reply:
I took this approach at first, but it turns out the `tok` is not a tal allocated array.  It's an offset into a tal allocated array. Its setup in  `parse_request` (jsonrpc.c:506).

Reply:
Ok you're right.  I tried a getting the tal_length of tok and it didn't work, so I bailed out. My bad.